pkg/controller/constraint/constraint_controller.go

```diff
@@ -609,35 +628,128 @@ func (c *ConstraintsCache) reportTotalConstraints(ctx context.Context, reporter
+	resList, err := clientset.Discovery().ServerResourcesForGroupVersion(groupVersion.String())
+	if err == nil {
+		for i := 0; i < len(resList.APIResources); i++ {
+			if resList.APIResources[i].Name == "validatingadmissionpolicies" {
+				VapAPIEnabled = new(bool)
+				*VapAPIEnabled = true
+				GroupVersion = &groupVersion
+				return true, GroupVersion
+			}
+		}
 	}
-	if _, err := clientset.Discovery().ServerResourcesForGroupVersion(groupVersion.String()); err != nil {
-		log.Info("IsVapAPIEnabled ServerResourcesForGroupVersion", "error", err)
-		VapAPIEnabled = new(bool)
-		*VapAPIEnabled = false
-		return false
+
+	groupVersion = schema.GroupVersion{Group: "admissionregistration.k8s.io", Version: "v1beta1"}
+	resList, err = clientset.Discovery().ServerResourcesForGroupVersion(groupVersion.String())
+	if err == nil {
+		for i := 0; i < len(resList.APIResources); i++ {
+			if resList.APIResources[i].Name == "validatingadmissionpolicies" {
+				VapAPIEnabled = new(bool)
+				*VapAPIEnabled = true
+				GroupVersion = &groupVersion
+				return true, GroupVersion
+			}
+		}
 	}
-	log.Info("IsVapAPIEnabled true")
+
+	log.Info("IsVapAPIEnabled", "error", err)
 	VapAPIEnabled = new(bool)
-	*VapAPIEnabled = true
-	return true
+	*VapAPIEnabled = false
+	return false, nil
+}
+
+func vapBindingForVersion(gvk schema.GroupVersion) (client.Object, error) {
+	switch gvk.Version {
+	case "v1":
+		return &admissionregistrationv1.ValidatingAdmissionPolicyBinding{}, nil
+	case "v1beta1":
+		return &admissionregistrationv1beta1.ValidatingAdmissionPolicyBinding{}, nil
+	default:
+		return nil, errors.New("unrecognized version")
+	}
+}
+
+func (r *ReconcileConstraint) getRunTimeVAPBinding(gvk *schema.GroupVersion, transformedVapBinding *admissionregistrationv1beta1.ValidatingAdmissionPolicyBinding, currentVapBinding client.Object) (client.Object, error) {
+	if currentVapBinding == nil {
+		if gvk.Version == "v1" {
+			return v1beta1ToV1(transformedVapBinding)
+		}
+		return transformedVapBinding.DeepCopy(), nil
+	}
+	if gvk.Version == "v1" {
+		v1CurrentVAPBinding, ok := currentVapBinding.(*admissionregistrationv1.ValidatingAdmissionPolicyBinding)
+		if !ok {
+			return nil, errors.New("Unable to convert to v1 VAP")
+		}
+		v1CurrentVAPBinding = v1CurrentVAPBinding.DeepCopy()
+		tempVAPBinding, err := v1beta1ToV1(transformedVapBinding)
+		if err != nil {
+			return nil, err
+		}
+		v1CurrentVAPBinding.Spec = tempVAPBinding.Spec
+		return v1CurrentVAPBinding, nil
+	}
+	v1beta1VAPBinding, ok := currentVapBinding.(*admissionregistrationv1beta1.ValidatingAdmissionPolicyBinding)
+	if !ok {
+		return nil, errors.New("Unable to convert to v1beta1 VAP")
+	}
+	v1beta1VAPBinding.Spec = transformedVapBinding.Spec
+	return v1beta1VAPBinding.DeepCopy(), nil
+}
+
+func v1beta1ToV1(v1beta1Obj *admissionregistrationv1beta1.ValidatingAdmissionPolicyBinding) (*admissionregistrationv1.ValidatingAdmissionPolicyBinding, error) {
```

Comment:
Can we not use scheme.Convert() here?

Reply:
Same happens with the binding as well, using `r.scheme.Convert` results in error in direct conversion.